pkg/metrics/garbagecollector.go

```diff
@@ -0,0 +1,183 @@
+	v1 "github.com/rancher/types/apis/core/v1"
+	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+var (
+	targetMetricsByNameForClientKey = []interface{}{
+		nm.TotalAddWS, nm.TotalRemoveWS, nm.TotalAddConnectionsForWS, nm.TotalRemoveConnectionsForWS,
+		nm.TotalTransmitBytesOnWS, nm.TotalTransmitErrorBytesOnWS, nm.TotalReceiveBytesOnWS,
+	}
+
+	targetMetricsByIPForPeer = []interface{}{
+		nm.TotalAddPeerAttempt, nm.TotalPeerConnected, nm.TotalPeerDisConnected,
+	}
+)
+
+type metricGarbageCollector struct {
+	clusterLister  v3.ClusterLister
+	nodeLister     v3.NodeLister
+	endpointLister v1.EndpointsLister
+}
+
+func (gc *metricGarbageCollector) metricGarbageCollection() {
+	logrus.Debugf("[metrics-garbage-collector] Start")
+
+	isClusterMode := settings.Namespace.Get() != "" && settings.PeerServices.Get() != ""
+
+	// Existing resources
+	observedResourceNames := map[string]bool{}
+	// Exisiting Metrics Labels
+	observedLabelsMap := map[string]map[interface{}][]map[string]string{}
+
+	// Get Clusters
+	clusters, err := gc.clusterLister.List("", labels.Everything())
+	if err != nil {
+		logrus.Errorf("[metrics-garbage-collector] failed to list clusters: %s", err)
+		return
+	}
+	for _, cluster := range clusters {
+		if _, ok := observedResourceNames[cluster.Name]; !ok {
+			observedResourceNames[cluster.Name] = true
+		}
+	}
+	// Get Nodes
+	nodes, err := gc.nodeLister.List("", labels.Everything())
+	if err != nil {
+		logrus.Errorf("[metrics-garbage-collector] failed to list nodes: %s", err)
+	}
+	for _, node := range nodes {
+		if _, ok := observedResourceNames[node.Namespace+":"+node.Name]; !ok {
+			observedResourceNames[node.Namespace+":"+node.Name] = true
+		}
+	}
+	// Get Endpoints
+	if isClusterMode {
+		endpoints, err := gc.endpointLister.List(settings.Namespace.Get(), labels.Everything())
+		if err != nil {
+			logrus.Errorf("[metrics-garbage-collector] failed to list endpoints: %s", err)
+		}
+		for _, svc := range strings.Split(settings.PeerServices.Get(), ",") {
+			for _, e := range endpoints {
+				// If Endpoint is associated with PeerServices service
+				if e.Name == strings.TrimSpace(svc) {
+					for _, subset := range e.Subsets {
+						for _, addr := range subset.Addresses {
+							observedResourceNames[addr.IP] = true
+						}
+					}
+				}
+			}
+		}
+	}
+
+	buildObservedLabelMaps(targetMetricsByNameForClientKey, "clientkey", observedLabelsMap)
+	buildObservedLabelMaps(targetMetricsByIPForPeer, "peer", observedLabelsMap)
+	buildObservedLabelMaps([]interface{}{clusterOwner}, "cluster", observedLabelsMap)
+
+	removedCount := removeMetricsForDeletedResource(observedLabelsMap, observedResourceNames)
+
+	logrus.Debugf("[metrics-garbage-collector] Finished - removed %d items", removedCount)
+}
+
+func buildObservedLabelMaps(collectors []interface{}, targetLabel string, observedLabels map[string]map[interface{}][]map[string]string) int {
+	// {
```

Comment:
do we need lines 95-102

Reply:
It's the structure of the maps this thing builds. I found it helpful...

Reply:
ah, may be we can add a line like: `Example of the map`. So the purpose of the commented stuff is more clear. What do you think

Reply:
Agreed. Updated